src/zod.ts

```diff
@@ -0,0 +1,81 @@
+import {
+  appendErrors,
+  FieldValues,
+  Resolver,
+  ResolverError,
+  ResolverSuccess,
+  transformToNestObject,
+} from 'react-hook-form';
+import { Schema, ZodError } from 'zod';
+import { ParseParams } from 'zod/lib/src/parser';
+import convertArrayToPathName from './utils/convertArrayToPathName';
+
+const parseErrorSchema = (
+  zodError: ZodError,
+  validateAllFieldCriteria: boolean,
+) => {
+  if (!Array.isArray(zodError.errors) || zodError.isEmpty) {
+    return {};
+  }
+
+  return zodError.errors.reduce<Record<string, any>>(
+    (previous, { path, message, code: type }) => {
+      const currentPath = convertArrayToPathName(path);
+
+      previous[currentPath] = {
+        message,
+        type,
+      };
+
+      return {
+        ...previous,
+        ...(path
+          ? previous[currentPath] && validateAllFieldCriteria
+            ? {
+                [currentPath]: appendErrors(
+                  currentPath,
+                  validateAllFieldCriteria,
+                  previous,
+                  type || '',
+                  message,
+                ),
+              }
+            : {
+                [currentPath]: previous[currentPath] || {
+                  message,
+                  type,
+                  ...(validateAllFieldCriteria
+                    ? {
+                        types: { [type || '']: message || true },
+                      }
+                    : {}),
+                },
+              }
+          : {}),
+      };
+    },
+    {},
+  );
+};
+
+export const zodResolver = <TFieldValues extends FieldValues>(
+  schema: Schema<any>,
+  options?: ParseParams,
+): Resolver<TFieldValues> => async (
```

Comment:
Otherwise type will not be infered from the Schema?

```suggestion
export const zodResolver = <TFieldValues extends FieldValues>(
  schema: Schema<TFieldValues>,
  options?: ParseParams,
): Resolver<TFieldValues> => async (
```

Reply:
```suggestion
export const zodResolver = <TFieldValues extends FieldValues, ZodTypeDef extends ZodTypeDef>(
  schema: Schema<TFieldValues, ZodTypeDef>,
  options?: ParseParams,
): Resolver<TFieldValues> => async (
```

actually would even be better, otherwise we couldn't pass a schema with a different type def

Reply:
Thanks for reviewing, it looks better :)